### PR TITLE
Experimental: precompile functions

### PR DESCRIPTION
**Pull request by IainNZ**
Test script:

```julia
start_t = time_ns()
println("start ", (time_ns()-start_t)/10^9); start_t = time_ns()
using JuMP
println("using ", (time_ns()-start_t)/10^9); start_t = time_ns()
m = Model()
println("model ", (time_ns()-start_t)/10^9); start_t = time_ns()
N=10000
@defVar(m, x[1:N])
println("defva ", (time_ns()-start_t)/10^9); start_t = time_ns()
@addConstraint(m, sum{i*x[i],i=1:N} >= N)
println("addcn ", (time_ns()-start_t)/10^9); start_t = time_ns()
```

### CODE REVIEW

**Comment by IainNZ** [comment]
Seems to help way more in global scope than in a function, not sure why.

**Comment by joehuchette** [comment]
I'm pretty sure the printing code takes forever to compile; what happens if you precompile some of that? Or use ``;`` in global scope...

**Comment by IainNZ** [comment]
Yeah printing takes forever, that was gonna my next challenge if this worked at all.

To be clear, I was running that code as a script (with precompiled JuMP), so its not doing any printing (I also run with coverage to identity what was running).

Most of the benefit came from the `addToExpression_reorder`, the rest from the `parse*` methods. The others didn't do anything as far as I can tell.

File: src/precompile.jl [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+Base.precompile(Variable, (Model,Float64,Float64,Symbol,String,Float64))
+Base.precompile(Variable, (Model,Float64,Int,Symbol,String,Float64))
+Base.precompile(Variable, (Model,Int,Float64,Symbol,String,Float64))
+Base.precompile(Variable, (Model,Int,Int,Symbol,String,Float64))
+Base.precompile(addConstraint, (Model,LinearConstraint))
+# macros.jl
+Base.precompile(buildrefsets, (Expr,))
+# parseExpr_staged.jl
+Base.precompile(addToExpression_reorder, (AffExpr,(Int,Variable)))
+Base.precompile(addToExpression_reorder, (AffExpr,(Float64,Int)))
```
Comment (joehuchette) [inline src/precompile.jl]: These are v0.4 only

**Comment by joehuchette** [comment]
Cool, looks like it's helping a lot with the constraint

File: src/precompile.jl (hunk 1 above)
Comment (IainNZ) [inline src/precompile.jl]: Yep

File: src/JuMP.jl [hunk 2]
```diff
@@ -1015,6 +1015,8 @@ include("callbacks.jl")
 include("print.jl")
 # Nonlinear-specific code
 include("nlp.jl")
+# Precompile functions to speed up initial runs
+include("precompile.jl")
```
Comment (mlubin) [inline src/JuMP.jl]: Conditional include on ``VERSION > v"0.4-"``?

Comment (IainNZ) [inline src/JuMP.jl]: If we merge this, yeah

File: src/precompile.jl [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+defva 0.325597146
+addcn 0.989031469
+=#
+
+# JuMP.jl
+Base.precompile(Variable, (Model,Float64,Float64,Symbol,String,Float64))
+Base.precompile(Variable, (Model,Float64,Int,Symbol,String,Float64))
+Base.precompile(Variable, (Model,Int,Float64,Symbol,String,Float64))
+Base.precompile(Variable, (Model,Int,Int,Symbol,String,Float64))
+Base.precompile(addConstraint, (Model,LinearConstraint))
```
Comment (tkelman) [inline src/precompile.jl]: can has nonlinear too?

**Comment by IainNZ** [comment]
```
@iaindunning: have you measured a benefit to doing that? unless you use it in sysimg/userimg, precompile doesn't actually save machine code.

slackbot [11:51 PM] 
spell-check "precompile": did you mean "magic"?

----- Today August 15th, 2015 -----

iaindunning [12:24 AM] 
@ihnorton: I'm seeing something

iaindunning [12:24 AM]
https://github.com/JuliaOpt/JuMP.jl/pull/531

GitHub
Experimental: precompile functions by IainNZ · Pull Request #531 · JuliaOpt/JuMP.jl · GitHub
Test script: start_t = time_ns() println("start ", (time_ns()-start_t)/10^9); start_t = time_ns() using JuMP println("using ", (time_ns()-start_t)/10^9); start_t = time_ns() m = Model() println("m...

iaindunning [12:24 AM]
i don't know how to explain it, if thats the case

iaindunning [12:24 AM]
maybe something to do with the generation function

ihnorton [12:29 AM] 
ah, yeah... with so much code generation that makes sense as precompiling should save you the fully expanded versions of everything

ihnorton [12:29 AM]
in lowered form

iaindunning [12:29 AM] 
the timing change was non-trivial, like 2x for the thing i was focused on

iaindunning [12:29 AM]
ahhh ok

iaindunning [12:29 AM]
so code with not involving macros would not likely see the benefit?

ihnorton [12:31 AM] 
jameson was talking about it earlier this evening and said (IIUC) that the benefit was minimal in the code he had profiled, but I think that was inference (which does a lot of things, but thankfully doesn't use @generated yet)
```

File: src/precompile.jl (hunk 3 above)
Comment (joehuchette) [inline src/precompile.jl]: @mlubin any idea what might help out with the nonlinear code?

Comment (mlubin) [inline src/precompile.jl]: Most of that is in ``ReverseDiffSparse``, which we shouldn't precompile here.

**Comment by IainNZ** [comment]
I was looking around and saw that PyCall was exploring similar ideas (https://github.com/stevengj/PyCall.jl/issues/167)

**Comment by IainNZ** [comment]
Closing in favour of #541

**Pull request closed without merging**